<a href="https://colab.research.google.com/github/RM559712/fase6_cap1/blob/main/EnydBentivoglio_rm560234_pbl_fase6.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
#Montar o ambiente no drive
from google.colab import drive
drive.mount('/content/drive')

MessageError: Error: credential propagation was unsuccessful

In [ ]:
#Clonar yolov5
! git clone https://github.com/ultralytics/yolov5.git

Cloning into 'yolov5'...
remote: Enumerating objects: 17410, done.
remote: Counting objects: 100% (84/84), done.
remote: Compressing objects: 100% (62/62), done.
remote: Total 17410 (delta 63), reused 22 (delta 22), pack-reused 17326 (from 3)
Receiving objects: 100% (17410/17410), 16.29 MiB | 18.50 MiB/s, done.
Resolving deltas: 100% (11928/11928), done.


In [ ]:
#Instalar os requerimentos
! pip install -r yolov5/requirements.txt

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 363.4/363.4 MB 3.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.8/13.8 MB 123.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.6/24.6 MB 84.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 883.7/883.7 kB 54.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 664.8/664.8 MB 1.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 211.5/211.5 MB 5.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.3/56.3 MB 15.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 127.9/127.9 MB 7.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 207.5/207.5 MB 4.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.1/21.1 MB 38.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 68.1 MB/s eta 0:00:00
  Attempting uninstall: nvidia-nvjitlink-cu12
    Found existing installation: nvidia-nvjitlink-cu12 12.5.82
    Uninstalling

In [ ]:
#Importar objects4.yaml
from google.colab import files
files.upload()

TypeError: 'NoneType' object is not subscriptable

In [ ]:
#Treinar o modelo v5 em 40 épocas
!python yolov5/train.py --data objects4.yaml --weights yolov5s.pt --img 640 --epochs 40

wandb: WARNING ⚠️ wandb is deprecated and will be removed in a future release. See supported integrations at https://github.com/ultralytics/yolov5#integrations.
2025-04-26 14:37:00.518348: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1745678220.539281    2296 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1745678220.545562    2296 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
wandb: Using wandb-core as the SDK backend.  Please refer to https://wandb.me/wandb-core for more information.
wandb: (1) Create a W&B account
wandb: (2) Use an existing W&B account
wandb: (3) Don't visualize my results
wandb: Enter your choice: (30 second timeout) 
wandb: W&

In [ ]:
#Salvar os resultados
!cp -r /content/yolov5/runs/train/exp2 /content/drive/MyDrive/my_yolo_results/

In [ ]:
#Teste do modelo v5 em 40 épocas
import os
import subprocess

def get_latest_train_run_folder():
    subfolders = [f.path for f in os.scandir('yolov5/runs/train') if f.is_dir()]
    latest_folder = max(subfolders, key=os.path.getctime, default=None)
    return latest_folder

latest_run = get_latest_train_run_folder()
result = subprocess.run(f'python yolov5/detect.py --weights {latest_run}/weights/best.pt --img 640 --source drive/MyDrive/FIAP/Teste/ --data yolov5/data/objects4.yaml', shell=True, capture_output=True, text=True)
if latest_run:
    # COMANDO
    result = subprocess.run(f'python yolov5/detect.py --weights {latest_run}/weights/best.pt --img 640 --source drive/MyDrive/FIAP/Teste/ --data yolov5/data/objects4.yaml', shell=True, capture_output=True, text=True)
    print(result.stdout)
    print(result.stderr)
else:
    print("Não foi possível encontrar a pasta de treinamento mais recente.")


detect: weights=['yolov5/runs/train/exp2/weights/best.pt'], source=drive/MyDrive/FIAP/Teste/, data=yolov5/data/objects4.yaml, imgsz=[640, 640], conf_thres=0.25, iou_thres=0.45, max_det=1000, device=, view_img=False, save_txt=False, save_format=0, save_csv=False, save_conf=False, save_crop=False, nosave=False, classes=None, agnostic_nms=False, augment=False, visualize=False, update=False, project=yolov5/runs/detect, name=exp, exist_ok=False, line_thickness=3, hide_labels=False, hide_conf=False, half=False, dnn=False, vid_stride=1
YOLOv5 🚀 v7.0-416-gfe1d4d99 Python-3.11.12 torch-2.6.0+cu124 CUDA:0 (Tesla T4, 15095MiB)

Fusing layers... 
Model summary: 157 layers, 7015519 parameters, 0 gradients, 15.8 GFLOPs
image 1/1 /content/drive/MyDrive/FIAP/Teste/val6.jpeg: 448x640 1 carro, 1 pessoa, 30.5ms
Speed: 0.6ms pre-process, 30.5ms inference, 128.7ms NMS per image at shape (1, 3, 640, 640)
Results saved to yolov5/runs/detect/exp2



In [ ]:
#Treinar o modelo v5 em 60 épocas
!python yolov5/train.py --data objects4.yaml --weights yolov5s.pt --img 640 --epochs 60

wandb: WARNING ⚠️ wandb is deprecated and will be removed in a future release. See supported integrations at https://github.com/ultralytics/yolov5#integrations.
2025-04-26 14:44:56.528778: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1745678696.563146    4420 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1745678696.573266    4420 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
wandb: Using wandb-core as the SDK backend.  Please refer to https://wandb.me/wandb-core for more information.
wandb: (1) Create a W&B account
wandb: (2) Use an existing W&B account
wandb: (3) Don't visualize my results
wandb: Enter your choice: (30 second timeout) 
wandb: W&

In [ ]:
#Salvar os resultados de treinamento do v5 em 60 épocas
!cp -r /content/yolov5/runs/train/exp3 /content/drive/MyDrive/my_yolo_results/

In [ ]:
#Testar modelo v5 treinado em 60 épocas
import os
import subprocess

def get_latest_train_run_folder():
    subfolders = [f.path for f in os.scandir('yolov5/runs/train') if f.is_dir()]
    latest_folder = max(subfolders, key=os.path.getctime, default=None)
    return latest_folder

latest_run = get_latest_train_run_folder()
result = subprocess.run(f'python yolov5/detect.py --weights {latest_run}/weights/best.pt --img 640 --source drive/MyDrive/FIAP/Teste/ --data yolov5/data/objects4.yaml', shell=True, capture_output=True, text=True)
if latest_run:
    # COMANDO
    result = subprocess.run(f'python yolov5/detect.py --weights {latest_run}/weights/best.pt --img 640 --source drive/MyDrive/FIAP/Teste/ --data yolov5/data/objects4.yaml', shell=True, capture_output=True, text=True)
    print(result.stdout)
    print(result.stderr)
else:
    print("Não foi possível encontrar a pasta de treinamento mais recente.")


detect: weights=['yolov5/runs/train/exp3/weights/best.pt'], source=drive/MyDrive/FIAP/Teste/, data=yolov5/data/objects4.yaml, imgsz=[640, 640], conf_thres=0.25, iou_thres=0.45, max_det=1000, device=, view_img=False, save_txt=False, save_format=0, save_csv=False, save_conf=False, save_crop=False, nosave=False, classes=None, agnostic_nms=False, augment=False, visualize=False, update=False, project=yolov5/runs/detect, name=exp, exist_ok=False, line_thickness=3, hide_labels=False, hide_conf=False, half=False, dnn=False, vid_stride=1
YOLOv5 🚀 v7.0-416-gfe1d4d99 Python-3.11.12 torch-2.6.0+cu124 CUDA:0 (Tesla T4, 15095MiB)

Fusing layers... 
Model summary: 157 layers, 7015519 parameters, 0 gradients, 15.8 GFLOPs
image 1/1 /content/drive/MyDrive/FIAP/Teste/val6.jpeg: 448x640 2 carros, 1 pessoa, 28.3ms
Speed: 0.5ms pre-process, 28.3ms inference, 117.9ms NMS per image at shape (1, 3, 640, 640)
Results saved to yolov5/runs/detect/exp4



In [ ]:
#Treinar modelo v5 em 80 épocas
!python yolov5/train.py --data objects4.yaml --weights yolov5s.pt --img 640 --epochs 80

wandb: WARNING ⚠️ wandb is deprecated and will be removed in a future release. See supported integrations at https://github.com/ultralytics/yolov5#integrations.
2025-04-26 14:52:34.836337: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1745679154.870416    6477 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1745679154.880599    6477 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
wandb: Using wandb-core as the SDK backend.  Please refer to https://wandb.me/wandb-core for more information.
wandb: (1) Create a W&B account
wandb: (2) Use an existing W&B account
wandb: (3) Don't visualize my results
wandb: Enter your choice: (30 second timeout) 
wandb: W&

In [ ]:
#Salvar ressultados v5 para 80 épocas
!cp -r /content/yolov5/runs/train/exp4 /content/drive/MyDrive/my_yolo_results/

In [ ]:
#Testar o modelo treinado em 80 épocas
import os
import subprocess

def get_latest_train_run_folder():
    subfolders = [f.path for f in os.scandir('yolov5/runs/train') if f.is_dir()]
    latest_folder = max(subfolders, key=os.path.getctime, default=None)
    return latest_folder

latest_run = get_latest_train_run_folder()
result = subprocess.run(f'python yolov5/detect.py --weights {latest_run}/weights/best.pt --img 640 --source drive/MyDrive/FIAP/Teste/ --data yolov5/data/objects4.yaml', shell=True, capture_output=True, text=True)
if latest_run:
    # COMANDO
    result = subprocess.run(f'python yolov5/detect.py --weights {latest_run}/weights/best.pt --img 640 --source drive/MyDrive/FIAP/Teste/ --data yolov5/data/objects4.yaml', shell=True, capture_output=True, text=True)
    print(result.stdout)
    print(result.stderr)
else:
    print("Não foi possível encontrar a pasta de treinamento mais recente.")


detect: weights=['yolov5/runs/train/exp4/weights/best.pt'], source=drive/MyDrive/FIAP/Teste/, data=yolov5/data/objects4.yaml, imgsz=[640, 640], conf_thres=0.25, iou_thres=0.45, max_det=1000, device=, view_img=False, save_txt=False, save_format=0, save_csv=False, save_conf=False, save_crop=False, nosave=False, classes=None, agnostic_nms=False, augment=False, visualize=False, update=False, project=yolov5/runs/detect, name=exp, exist_ok=False, line_thickness=3, hide_labels=False, hide_conf=False, half=False, dnn=False, vid_stride=1
YOLOv5 🚀 v7.0-416-gfe1d4d99 Python-3.11.12 torch-2.6.0+cu124 CUDA:0 (Tesla T4, 15095MiB)

Fusing layers... 
Model summary: 157 layers, 7015519 parameters, 0 gradients, 15.8 GFLOPs
image 1/1 /content/drive/MyDrive/FIAP/Teste/val6.jpeg: 448x640 2 carros, 1 pessoa, 28.8ms
Speed: 0.6ms pre-process, 28.8ms inference, 122.0ms NMS per image at shape (1, 3, 640, 640)
Results saved to yolov5/runs/detect/exp6



In [ ]:
#Clonar Yolov8
!git clone https://github.com/ultralytics/ultralytics.git

Cloning into 'ultralytics'...
remote: Enumerating objects: 57192, done.
remote: Counting objects: 100% (834/834), done.
remote: Compressing objects: 100% (564/564), done.
remote: Total 57192 (delta 636), reused 277 (delta 270), pack-reused 56358 (from 3)
Receiving objects: 100% (57192/57192), 32.35 MiB | 27.35 MiB/s, done.
Resolving deltas: 100% (42221/42221), done.


In [ ]:
#Yolov8 requerimentos
!pip install ultralytics --upgrade

In [ ]:
#Treinar v8 em 40 épocas
!yolo task=detect mode=train data=objects4.yaml model=yolov8n.pt epochs=40 imgsz=640

100% 6.25M/6.25M [00:00<00:00, 100MB/s]
Ultralytics 8.3.117 🚀 Python-3.11.12 torch-2.6.0+cu124 CUDA:0 (Tesla T4, 15095MiB)
engine/trainer: task=detect, mode=train, model=yolov8n.pt, data=objects4.yaml, epochs=40, time=None, patience=100, batch=16, imgsz=640, save=True, save_period=-1, cache=False, device=None, workers=8, project=None, name=train, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, embed=None, show=False, save_frames=False, save_txt=False, save_conf=False, save_crop=False, show_labels=True, show_conf=True, s

In [ ]:
#Salvar treinamento de v8 em 40 épocas
!cp -r /content/runs/detect/train /content/drive/MyDrive/YOLO_results_v8/

In [ ]:
#Testar treinamento de v8 em 40 épocas
import os
from ultralytics import YOLO

def get_latest_train_run_folder():
    subfolders = [f.path for f in os.scandir('/content/runs/detect') if f.is_dir()]
    latest_folder = max(subfolders, key=os.path.getctime, default=None)
    return latest_folder

latest_run = get_latest_train_run_folder()

if latest_run:
    model = YOLO(f'{latest_run}/weights/best.pt')  # Load the best model
    results = model.predict(source='/content/drive/MyDrive/FIAP/Teste', imgsz=640, save=True, conf=0.25)
else:
    print("Não foi possível encontrar a pasta de treinamento mais recente.")



image 1/1 /content/drive/MyDrive/FIAP/Teste/val6.jpeg: 448x640 (no detections), 41.2ms
Speed: 3.9ms preprocess, 41.2ms inference, 31.1ms postprocess per image at shape (1, 3, 448, 640)
Results saved to runs/detect/predict


In [ ]:
#Treinamento de v8 em 60 épocas
!yolo task=detect mode=train data=objects4.yaml model=yolov8n.pt epochs=60 imgsz=640

Ultralytics 8.3.117 🚀 Python-3.11.12 torch-2.6.0+cu124 CUDA:0 (Tesla T4, 15095MiB)
engine/trainer: task=detect, mode=train, model=yolov8n.pt, data=objects4.yaml, epochs=60, time=None, patience=100, batch=16, imgsz=640, save=True, save_period=-1, cache=False, device=None, workers=8, project=None, name=train2, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, embed=None, show=False, save_frames=False, save_txt=False, save_conf=False, save_crop=False, show_labels=True, show_conf=True, show_boxes=True, line_width=None, format

In [ ]:
#Salvar treinamento de v8 em 60 épocas
!cp -r /content/runs/detect/train2/ /content/drive/MyDrive/YOLO_results_v8/

In [ ]:
#Listar se os pesos estao okay
!ls /content/runs/detect/train*/weights/

/content/runs/detect/train2/weights/:
best.pt  last.pt

/content/runs/detect/train/weights/:
best.pt  last.pt


In [ ]:
#Testar treinamento de v8 em 60 épocas
import os
from ultralytics import YOLO

def get_latest_train_run_folder():
    subfolders = [f.path for f in os.scandir('/content/runs/detect') if f.is_dir() and os.path.basename(f.path).startswith('train')]
    latest_folder = max(subfolders, key=os.path.getctime, default=None)
    return latest_folder

latest_run = get_latest_train_run_folder()

if latest_run:
    weights_path = f'{latest_run}/weights/best.pt'
    if not os.path.exists(weights_path):
        print(f"{weights_path} not found, trying last.pt instead.")
        weights_path = f'{latest_run}/weights/last.pt'

    model = YOLO(weights_path)  # Load best.pt or fallback to last.pt
    results = model.predict(source='/content/drive/MyDrive/FIAP/Teste', imgsz=640, save=True, conf=0.25)
else:
    print("Não foi possível encontrar a pasta de treinamento mais recente.")



image 1/1 /content/drive/MyDrive/FIAP/Teste/val6.jpeg: 448x640 (no detections), 7.1ms
Speed: 2.7ms preprocess, 7.1ms inference, 0.8ms postprocess per image at shape (1, 3, 448, 640)
Results saved to runs/detect/predict3


In [ ]:
#Treinamento de v8 em 80 épocas
!yolo task=detect mode=train data=objects4.yaml model=yolov8n.pt epochs=80 imgsz=640

Ultralytics 8.3.117 🚀 Python-3.11.12 torch-2.6.0+cu124 CUDA:0 (Tesla T4, 15095MiB)
engine/trainer: task=detect, mode=train, model=yolov8n.pt, data=objects4.yaml, epochs=80, time=None, patience=100, batch=16, imgsz=640, save=True, save_period=-1, cache=False, device=None, workers=8, project=None, name=train3, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, embed=None, show=False, save_frames=False, save_txt=False, save_conf=False, save_crop=False, show_labels=True, show_conf=True, show_boxes=True, line_width=None, format

In [ ]:
#Salvar treinamento de v8 em 80 épocas
!cp -r /content/runs/detect/train3/ /content/drive/MyDrive/YOLO_results_v8/

In [ ]:
#Testar treinamento de v8 em 80 épocas
import os
from ultralytics import YOLO

def get_latest_train_run_folder():
    subfolders = [f.path for f in os.scandir('/content/runs/detect') if f.is_dir() and os.path.basename(f.path).startswith('train')]
    latest_folder = max(subfolders, key=os.path.getctime, default=None)
    return latest_folder

latest_run = get_latest_train_run_folder()

if latest_run:
    weights_path = f'{latest_run}/weights/best.pt'
    if not os.path.exists(weights_path):
        print(f"{weights_path} not found, trying last.pt instead.")
        weights_path = f'{latest_run}/weights/last.pt'

    model = YOLO(weights_path)  # Load best.pt or fallback to last.pt
    results = model.predict(source='/content/drive/MyDrive/FIAP/Teste', imgsz=640, save=True, conf=0.25)
else:
    print("Não foi possível encontrar a pasta de treinamento mais recente.")



image 1/1 /content/drive/MyDrive/FIAP/Teste/val6.jpeg: 448x640 3 pessoas, 7.8ms
Speed: 2.6ms preprocess, 7.8ms inference, 121.8ms postprocess per image at shape (1, 3, 448, 640)
Results saved to runs/detect/predict4


###Comparação Unificada: YOLOv5 e YOLOv8 (60 e 80 épocas)

---

### Facilidade de Uso/Integração

- *YOLOv5 (60 e 80 épocas)*:
  - Fácil de usar, altamente compatível com Python e Google Colab.
  - Fácil ajuste de configurações e visualização.

- *YOLOv8 (60 e 80 épocas)*:
  - Interface mais moderna e simplificada.
  - Integração direta com Ultralytics Hub

*Conclusão: Ambos são simples, mas o v8 se destaca por ser mais fluido

---

###  Precisão do Modelo (mAP@0.5)

| Modelo           | mAP@0.5 |
|------------------|---------|
| YOLOv5 (60 ep)   | 0.656   |
| YOLOv5 (80 ep)   | 0.630   |
| YOLOv8 (60 ep)   | 0.720   |
| YOLOv8 (80 ep)   | 0.850   |

*Conclusão: Apesar do O YOLOv5 com 60 épocas surpreender positivamente, o YOLOv8 com 80 épocas lidera com folga.

---

### Tempo de Treinamento

| Modelo           | Total estimado |
|------------------|----------------|
| YOLOv5 (60 ep)   | ~3.6 min       |
| YOLOv5 (80 ep)   | ~4.8 min       |
| YOLOv8 (60 ep)   | ~9.0 min       |
| YOLOv8 (80 ep)   | ~6.5 min       |

*Conclusão: O YOLOv5 tem treinamento significativamente mais rápido.

---

### ⚡ Tempo de Inferência

| Modelo           | Tempo (ms/imagem)  |
|------------------|--------------------|
| YOLOv5 (60 ep)   | ~15–25             |
| YOLOv5 (80 ep)   | ~15–25             |
| YOLOv8 (60 ep)   | ~20–35             |
| YOLOv8 (80 ep)   | ~20–35             |

 *Conclusão: Todos são rápidos, mas o YOLOv5 é mais leve.

---

### Conclusão Geral

Para melhores resultados de precisão, o YOLOv8 com 80 épocas é a melhor escolha. Se o objetivo for leveza e velocidade em dispositivos embarcados, o YOLOv5 leva vantagem em qualquer uma das épocas testadas.